### ISSUE TITLE
Why cant decorator parametrize automatically get function's arguments?

### ISSUE THREAD

**Issue by nitrocode** [pytest-dev/pytest#1742]
I don't understand why a pytest's test function's params have to be manually kept consistent using the pytest.mark.parametrize function.

Isn't there a way for pytest to automatically grab the arguments `("fileName", "fields", "thresh", "companies")` and assume that the fields provided `(fileNameA, fileNameAFields, 5, False)` are in the correct order to feed it into `test_me()` ?

```python
    @pytest.mark.parametrize(
        ("fileName", "fields", "thresh", "companies"), [ # this line should be automatic
            (fileNameA, fileNameAFields, 5, False),
            (fileNameB, fileNameBFields, 5, True),
        ],
        ids=[
            'FileA',
            'FileB'
        ]
    )
    def test_me(self, fileName, fields, thresh, companies):
    ...
```

**Comment by The-Compiler** [comment]
You can also mix fixtures and parametrization - sure, pytest could just assume everything which isn't a fixture is parametrized, but that sounds error-prone.

**Comment by RonnyPfannschmidt** [comment]
the parameterize mark is independent of the test function, it could also be used in a global mark, or in a class level mark

**Comment by nicoddemus** [comment]
A quick example of the latter, from pytest's own test suite:

```python
@pytest.mark.parametrize('flavor', ['fixture', 'yield_fixture'])
class TestContextManagerFixtureFuncs:

    def test_simple(self, testdir, flavor):
        ...

    def test_scoped(self, testdir, flavor):
        ...
```

**Comment by mgeier** [comment]
@nitrocode I know I'm quite late with this, but probably you (or somebody else who stumbles upon this) might still find this useful.

I totally agree that the argument names should be deduced automatically. I asked for this in #518 and I suggested an implementation in #780.

Sadly, this wasn't accepted, but I got some help for implementing this as an extension:

https://github.com/mgeier/pytest-auto-parametrize/
https://pypi.python.org/pypi/pytest-auto-parametrize/

I still hope that it (or something similar) gets included in core `pytest` at some point, because as a plugin it has around zero visibility, while if it were mentioned in the docs, people would actually know about this feature ...
